src/z_data/z_store.rs

```diff
@@ -0,0 +1,180 @@
+#[cfg(not(target_arch = "wasm32"))]
+use proptest::prelude::*;
+#[cfg(not(target_arch = "wasm32"))]
+use proptest_derive::Arbitrary;
+use serde::{Deserialize, Serialize};
+
+use std::collections::BTreeMap;
+
+use crate::hash::PoseidonCache;
+use crate::ptr::Ptr;
+use crate::store::Store;
+use crate::symbol::Symbol;
+use crate::tag::ExprTag;
+use crate::uint::UInt;
+use crate::z_cont::ZCont;
+use crate::z_expr::ZExpr;
+use crate::z_ptr::ZContPtr;
+use crate::z_ptr::ZExprPtr;
+use crate::z_ptr::ZPtr;
+
+use crate::field::LurkField;
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone, Default)]
+#[cfg_attr(not(target_arch = "wasm32"), derive(Arbitrary))]
+#[cfg_attr(not(target_arch = "wasm32"), proptest(no_bound))]
+pub struct ZStore<F: LurkField> {
+    pub expr_map: BTreeMap<ZExprPtr<F>, Option<ZExpr<F>>>,
+    pub cont_map: BTreeMap<ZContPtr<F>, Option<ZCont<F>>>,
+}
+
+impl<F: LurkField> ZStore<F> {
+    pub fn new() -> Self {
+        ZStore {
+            expr_map: BTreeMap::new(),
+            cont_map: BTreeMap::new(),
+        }
+    }
+
+    // Convert entire store to ZStore
+    pub fn to_z_store(store: &mut Store<F>) -> Self {
+        store.hydrate_scalar_cache();
+        let mut zstore = ZStore::new();
+        for zptr in store.z_expr_ptr_map.keys_cloned() {
+            let ptr = store.z_expr_ptr_map.get(&zptr).unwrap();
+            let zexpr = store.to_z_expr(ptr);
+            zstore.expr_map.insert(zptr, zexpr);
+        }
+        for zptr in store.z_cont_ptr_map.keys_cloned() {
+            let ptr = store.z_cont_ptr_map.get(&zptr).unwrap();
+            let zcont = store.to_z_cont(ptr);
+            zstore.cont_map.insert(zptr, zcont);
+        }
```

Comment:
I think implementing additional infrastructure on `CacheMap` is something that should happen in a future "figure out the right CacheMap represenation" PR.

In this particular case we have to clone/copy the the keys in the `CacheMap` anyway so that we can insert them into the ZStore so using `keys_cloned` shouldn't cost anything extra, and the call to `get` is safe given that we never delete values.